metricbeat/scripts/mage/docsCollector.go

```diff
@@ -0,0 +1,488 @@
+		return moduleData{}, errors.Wrap(err, "failed to read from spec file")
+	}
+	//Cheat and use the same struct.
+	var mod []moduleData
+	if err = yaml.Unmarshal(fd, &mod); err != nil {
+		return mod[0], err
+	}
+	module := mod[0]
+
+	rel, err := getRelease(module.Release)
+	if err != nil {
+		return mod[0], err
+	}
+	module.Release = rel
+
+	return module, nil
+}
+
+//The doc generator only needs the release field from the metricset-level fields.yml.
+func getReleaseState(metricsetPath string) (string, error) {
+	raw, err := ioutil.ReadFile(metricsetPath)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read from spec file")
+	}
+
+	type metricset struct {
+		Release string `yaml:"release"`
+	}
+	var rel []metricset
+	if err = yaml.Unmarshal(raw, &rel); err != nil {
+		return "", err
+	}
+
+	return getRelease(rel[0].Release)
+}
+
+//check to see if the metricset has dashboards
+func hasDashboards(modulePath string) bool {
+	_, err := os.Stat(filepath.Join(modulePath, "_meta/kibana"))
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+//use the reference file if it exists. if not, the normal one.
+func getConfigfile(modulePath string) (string, error) {
+	var cfgFile string
+	cfg := filepath.Join(modulePath, "_meta/config.reference.yml")
+	_, err := os.Stat(cfg)
+	if err == nil {
+		cfgFile = cfg
+	} else if os.IsNotExist(err) {
+		cfgBack := filepath.Join(modulePath, "_meta/config.yml")
+		if _, err := os.Stat(cfgBack); err != nil {
+			return "", fmt.Errorf("config file in %s could not be found", modulePath)
+		}
+		cfgFile = cfgBack
+	} else {
+		return "", fmt.Errorf("config file in %s could not be found", modulePath)
+	}
+
+	raw, err := ioutil.ReadFile(cfgFile)
+	return string(raw), err
+
+}
+
+//gather all the metricsets for a given module
+func gatherMetricsets(modulePath string, moduleName string) ([]metricsetData, error) {
+	metricsetList, err := filepath.Glob(filepath.Join(modulePath, "/*"))
+	if err != nil {
+		return nil, err
+	}
+	var metricsets []metricsetData
+	for _, metricset := range metricsetList {
+		isMetricset, err := testIfDocsInDir(metricset)
+		if err != nil {
+			return nil, err
+		}
+		if !isMetricset {
+			continue
+		}
+		metricsetName := filepath.Base(metricset)
+		release, err := getReleaseState(filepath.Join(metricset, "_meta/fields.yml"))
+		if err != nil {
+			return nil, err
+		}
+
+		//generate the asciidoc link used in the module docs, since we need this in a few places
+		link := fmt.Sprintf("<<metricbeat-metricset-%s-%s,%s>>", moduleName, metricsetName, metricsetName)
+
+		//test to see if the metricset has a data.json
+		hasData := false
+		_, err = os.Stat(filepath.Join(metricset, "_meta/data.json"))
+		if err == nil {
+			hasData = true
+		}
+
+		ms := metricsetData{
+			Path:       metricset,
+			Title:      metricsetName,
+			Release:    release,
+			Link:       link,
+			DataExists: hasData,
+		}
+
+		metricsets = append(metricsets, ms)
+
+	} // end of metricset loop
```

Comment:
If you feel the necessity of documenting a long loop maybe another option is to make the body a function, something like:
```
for _, metricset := range metricsetList {
		isMetricset, err := testIfDocsInDir(metricset)
		if err != nil {
			return nil, err
		}
		if !isMetricset {
			continue
		}
                ms, err := gatherMetricSet(moduleName, metricset)
                if err != nil {
                    return nil, err
                }
                metricsets = append(metricsets, ms)
}
```

Reply:
Yah, I tend to do that more out of habit than anything.